### PR TITLE
Add parser for Optional types to avoid the null madness

### PR DESCRIPTION
**Pull request by lucidd**
This adds a parser for `Optional<T>` for all types `T` we can already parse.
The main motivation behind this is to banish `null` as much as possible in favor of the much saver abstraction `Optional<T>` which denotes a possibly absent value. This addresses concerns raised in #77

### CODE REVIEW

**Comment by lucidd** [comment]
I just rebased this to the latest master. I would say we merge this now as is even if we are not able to completely enforce that non optional values are never null its still valuable to mark `Optional` values as such and have them parsed accordingly.

**Comment by renner** [comment]
We can merge this if you need it, choose yourself if you want to merge it from here or include the commits with #109.

File: src/test/java/com/suse/saltstack/netapi/parser/JsonParserTest.java [hunk 1]
```diff
@@ -202,6 +206,41 @@ public void testSaltStackJobsWithKwargsParser() throws Exception {
...
+        List<Optional<Integer>> expected = new LinkedList<>();
+        expected.add(Optional.of(1));
+        expected.add(Optional.of(2));
+        expected.add(Optional.of(3));
+        expected.add(Optional.empty());
+        expected.add(Optional.of(5));
+        assertEquals(expected, result.maybeInts);
+    }
+
+
```
Comment (renner) [inline src/test/java/com/suse/saltstack/netapi/parser/JsonParserTest.java]: Let's put only one empty line between methods.

File: src/test/java/com/suse/saltstack/netapi/parser/OptionalTest.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Helper Type to test the Optional Parser
+ */
+public class OptionalTest {
+
+    public Optional<String> nullString = Optional.empty();
+    public Optional<String> valueString = Optional.empty();
+    public Optional<String> absentString = Optional.empty();
+    public List<Optional<Integer>> maybeInts;
+
```
Comment (renner) [inline src/test/java/com/suse/saltstack/netapi/parser/OptionalTest.java]: Let's get also rid of this one.

File: src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java [hunk 3]
```diff
@@ -127,6 +131,45 @@ public Date read(JsonReader jsonReader) throws IOException {
...
+                    if (in.peek() == JsonToken.NULL) {
+                        in.nextNull();
+                        return Optional.empty();
+                    } else {
+                        A value = innerAdapter.read(in);
+                        return Optional.of(value);
+                    }
+                }
+
+                public void write(JsonWriter out, Optional<A> optional) throws IOException {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java]: Aren't `read()` and `write()` overrides from `TapeAdapter`? Let's put the `@Override` annotation?

File: src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java [hunk 4]
```diff
@@ -127,6 +131,45 @@ public Date read(JsonReader jsonReader) throws IOException {
...
+                    innerAdapter.write(out, optional.orElse(null));
+                }
+            };
+        }
+
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java]: Please remove this one as well.

File: src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java [hunk 5]
```diff
@@ -127,6 +131,45 @@ public Date read(JsonReader jsonReader) throws IOException {
     }
 
     /**
+     * TypeAdaptorFactory creating TypeAdapters for Optional
+     */
+    private class OptionalTypeAdapterFactory implements TypeAdapterFactory {
+
+        @SuppressWarnings("unchecked")
+        public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> typeToken) {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java]: Public methods should generally have javadoc, consider to add some documentation.

Comment (lucidd) [inline src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java]: create is overridden from `TypeAdapterFactory` so its already documented there i put the `@Override` on it to make it clear.

**Comment by lucidd** [comment]
I would prefer merging this PR first i will rebase #109 accordingly.

**Pull request merged**